tests/test_notifier.py

```diff
@@ -0,0 +1,341 @@
+                                 sqs_polling_timeout=5,
+                                 timeout=self.timeout,
+                                 overhead=self.overhead)
+        self.notifier.deploy()
+
+    def _mock_context(self, total_attempts):
+        latency = sum(delay for delay in self.delays)
+        average_overhead = self.overhead / 3  # FIXME: this is a guess
+        average_timeout = self.timeout / 2  # FIXME: this is a guess
+        parallelism = self.num_queues + (self.num_workers - self.num_queues) / 5  # FIXME: this is a guess
+        total_time = latency + (average_timeout + average_overhead) * total_attempts / parallelism
+        return MockLambdaContext(total_time)
+
+    def tearDown(self):
+        self.notifier.destroy()
+
+    def test(self):
+        expected_receptions = set()
+        expected_misses = set()
+        total_attempts = 0
+
+        def notify(expect: bool,  # whether the message should make it
+                   max_attempts: Optional[int] = None,  # how many attempts to allow
+                   responses: List[Tuple[float, int]] = None,  # a list of (delay, http_status) tuples, one per attempt
+                   attempts=None):  # expected number of attempts, currently only used to estmate the running time
+            if responses is None:
+                responses = [(0.0, 200)]
+            verify = random.random() > .5
+            notification_id = str(next(self.notification_id))
+            body = dict(notification_id=notification_id,
+                        responses=responses,
+                        verify=verify)
+            notification = Notification.from_scratch(notification_id=notification_id,
+                                                     subscription_id=str(random.choice(self.subscription_ids)),
+                                                     url=f"http://{self.address}:{self.port}/{notification_id}",
+                                                     method='POST',
+                                                     encoding='application/json',
+                                                     body=body,
+                                                     attempts=max_attempts,
+                                                     hmac_key=PostTestHandler.hmac_secret_key if verify else None,
+                                                     hmac_key_id='1234' if verify else None)
+            nonlocal total_attempts
+            total_attempts += min(notification.attempts, self.num_queues) if attempts is None else attempts
+            (expected_receptions if expect else expected_misses).add(notification_id)
+            self.notifier.enqueue(notification)
+
+        for repeat in range(self.repeats):
+            # A notification with …
+            # … zero permitted attempts will be missed.
+            notify(responses=[], max_attempts=0, expect=False)
+            for n in range(1, self.num_queues + 1):
```

Comment:
unless all of these test are dependent on each other, I suggest you make use of `subtest`. It will make debugging easier.

Reply:
They are dependent. Due to the non-deterministic timing of the code under test this test sends off all notifications first and then asserts their reception at the end, outside the loop. A `subtest` without assertions in its body would not make sense here. A failed assertion includes the ID of a missed or duplicate reception so it can be traced back to the line where it was sent.